source/common/http/http1/codec_impl.cc

```diff
@@ -534,25 +536,31 @@ int ConnectionImpl::onHeadersCompleteBase() {
     }
   }
 
-  int rc = onHeadersComplete(std::move(current_header_map_));
+  seen_content_length_ = current_header_map_->ContentLength() != nullptr;
+
+  const int rc = onHeadersComplete(std::move(current_header_map_));
   current_header_map_.reset();
   header_parsing_state_ = HeaderParsingState::Done;
 
-  // Returning 2 informs http_parser to not expect a body or further data on this connection.
+  // Returning 2 informs llhttp to not expect a body or further data on this connection.
   return handling_upgrade_ ? 2 : rc;
 }
 
-void ConnectionImpl::onMessageCompleteBase() {
+int ConnectionImpl::onMessageCompleteBase() {
   ENVOY_CONN_LOG(trace, "message complete", connection_);
   if (handling_upgrade_) {
     // If this is an upgrade request, swallow the onMessageComplete. The
     // upgrade payload will be treated as stream body.
     ASSERT(!deferred_end_stream_headers_);
     ENVOY_CONN_LOG(trace, "Pausing parser due to upgrade.", connection_);
+#ifdef ENVOY_ENABLE_LEGACY_HTTP_PARSER
```

Comment:
can we always return HPE_PAUSED and let the calling site call http_parser_pause? i.e. here: https://github.com/envoyproxy/envoy/pull/9033/files#diff-2ba39613a1b78b994ec31ddd03cf42f5R342

same for other places.